File: scripts/scan/structure.py
"""F-036 — top-level directory shape + entity-candidate file detection.

Pure read-only walk capped at depth 3 from ``root``. Skips a fixed list of
build / VCS / vendored dirs to keep the LLM input budget bounded.
"""

from __future__ import annotations

from pathlib import Path
from typing import Optional


_IGNORED_DIRS = frozenset({
    ".git",
    ".hg",
    ".svn",
    "node_modules",
    "__pycache__",
    ".venv",
    "venv",
    "env",
    "dist",
    "build",
    "target",
    ".harness",
    ".claude",
    ".pytest_cache",
    ".mypy_cache",
    ".ruff_cache",
    ".idea",
    ".vscode",
})

_ADR_CANDIDATES = ("docs/adr", "docs/decisions", "adr", "decisions")
_README_CANDIDATES = ("README.md", "README.rst", "README.txt", "readme.md")
_ENTITY_FILE_NAMES = frozenset({"models.py", "schemas.py", "entities.py"})
_ENTITY_DIR_NAMES = frozenset({"models", "schemas", "entities", "domain"})

_MAX_ENTITY_FILES = 24


def scan_structure(root: Path) -> dict:
    """Return a structure summary for ``root``.

    Keys:
        top_dirs (list[str]): immediate subdirectories, sorted, ignoring noise.
        adr_dir (str | None): relative path to a detected ADR directory.
        entity_candidate_files (list[str]): up to 24 relative paths likely to
            contain domain models (capped to bound LLM input).
        readme_path (str | None): relative path to the project README.
    """
    root = Path(root)
    if not root.is_dir():
        return {
            "top_dirs": [],
            "adr_dir": None,
            "entity_candidate_files": [],
            "readme_path": None,
        }

    top_dirs = sorted(
        child.name
        for child in root.iterdir()
        if child.is_dir() and child.name not in _IGNORED_DIRS and not child.name.startswith(".")
    )

    adr_dir = _find_adr_dir(root)
    readme = _find_readme(root)
    entities = _find_entity_candidates(root)

    return {
        "top_dirs": top_dirs,
        "adr_dir": adr_dir,
        "entity_candidate_files": entities,
        "readme_path": readme,
    }


def _find_adr_dir(root: Path) -> Optional[str]:
    for candidate in _ADR_CANDIDATES:
        if (root / candidate).is_dir():
            return candidate
    return None


def _find_readme(root: Path) -> Optional[str]:
    for candidate in _README_CANDIDATES:
        if (root / candidate).is_file():
            return candidate
    return None


def _find_entity_candidates(root: Path) -> list[str]:
    found: list[str] = []
    for path in _walk(root, max_depth=3):
        rel = path.relative_to(root).as_posix()
        if path.is_file():
            if path.name in _ENTITY_FILE_NAMES:
                found.append(rel)
            elif path.suffix == ".ts" and ".entity." in path.name:
                found.append(rel)
        elif path.is_dir() and path.name in _ENTITY_DIR_NAMES:
            for child in sorted(path.rglob("*")):
                if child.is_file() and child.suffix in (".py", ".ts", ".js"):
                    found.append(child.relative_to(root).as_posix())
                if len(found) >= _MAX_ENTITY_FILES:
                    break
        if len(found) >= _MAX_ENTITY_FILES:
            break
    return found[:_MAX_ENTITY_FILES]


def _walk(root: Path, *, max_depth: int):
    stack = [(root, 0)]
    while stack:
        current, depth = stack.pop()
        if depth > max_depth:
            continue
        try:
            children = sorted(current.iterdir())
        except (PermissionError, FileNotFoundError):
            continue
        for child in children:
            if child.name in _IGNORED_DIRS:
                continue
            if child.is_symlink():
                continue
            yield child
            if child.is_dir():
                stack.append((child, depth + 1))

File: src/scan/structure.ts
/**
 * Top-level directory shape + entity-candidate file detection (F-101
 * port of `scripts/scan/structure.py`, F-036).
 *
 * Pure read-only walk capped at depth 3 from `root`. Skips a fixed
 * list of build / VCS / vendored dirs to keep the LLM input budget
 * bounded.
 *
 * @module scan/structure
 */

import {readdirSync, statSync} from 'node:fs';
import {join, posix as pathPosix, relative} from 'node:path';

const IGNORED_DIRS: ReadonlySet<string> = new Set([
  '.git',
  '.hg',
  '.svn',
  'node_modules',
  '__pycache__',
  '.venv',
  'venv',
  'env',
  'dist',
  'build',
  'target',
  '.harness',
  '.claude',
  '.pytest_cache',
  '.mypy_cache',
  '.ruff_cache',
  '.idea',
  '.vscode',
]);

const ADR_CANDIDATES: ReadonlyArray<string> = [
  'docs/adr',
  'docs/decisions',
  'adr',
  'decisions',
];
const README_CANDIDATES: ReadonlyArray<string> = [
  'README.md',
  'README.rst',
  'README.txt',
  'readme.md',
];
const ENTITY_FILE_NAMES: ReadonlySet<string> = new Set(['models.py', 'schemas.py', 'entities.py']);
const ENTITY_DIR_NAMES: ReadonlySet<string> = new Set(['models', 'schemas', 'entities', 'domain']);
const MAX_ENTITY_FILES = 24;

/** Outcome of a structure scan. */
export interface StructureSummary {
  top_dirs: string[];
  adr_dir: string | null;
  entity_candidate_files: string[];
  readme_path: string | null;
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function isSymlink(path: string): boolean {
  try {
    return statSync(path).isSymbolicLink();
  } catch {
    return false;
  }
}

function toPosix(p: string): string {
  return p.split(/[\\/]/).filter((s) => s.length > 0).join('/');
}

/** Returns the directory shape summary for `root`. */
export function scanStructure(root: string): StructureSummary {
  if (!isDirectory(root)) {
    return {top_dirs: [], adr_dir: null, entity_candidate_files: [], readme_path: null};
  }

  let entries: string[];
  try {
    entries = readdirSync(root);
  } catch {
    entries = [];
  }
  const topDirs: string[] = [];
  for (const name of entries) {
    if (IGNORED_DIRS.has(name) || name.startsWith('.')) {
      continue;
    }
    if (isDirectory(join(root, name))) {
      topDirs.push(name);
    }
  }
  topDirs.sort();

  return {
    top_dirs: topDirs,
    adr_dir: findAdrDir(root),
    entity_candidate_files: findEntityCandidates(root),
    readme_path: findReadme(root),
  };
}

function findAdrDir(root: string): string | null {
  for (const candidate of ADR_CANDIDATES) {
    if (isDirectory(join(root, candidate))) {
      return candidate;
    }
  }
  return null;
}

function findReadme(root: string): string | null {
  for (const candidate of README_CANDIDATES) {
    if (isFile(join(root, candidate))) {
      return candidate;
    }
  }
  return null;
}

function findEntityCandidates(root: string): string[] {
  const found: string[] = [];
  for (const path of walk(root, 3)) {
    if (found.length >= MAX_ENTITY_FILES) {
      break;
    }
    const rel = toPosix(relative(root, path));
    if (isFile(path)) {
      const name = path.split(/[\\/]/).pop()!;
      if (ENTITY_FILE_NAMES.has(name)) {
        found.push(rel);
      } else if (name.endsWith('.ts') && name.includes('.entity.')) {
        found.push(rel);
      }
    } else if (isDirectory(path)) {
      const name = path.split(/[\\/]/).pop()!;
      if (ENTITY_DIR_NAMES.has(name)) {
        for (const child of walkAll(path)) {
          if (found.length >= MAX_ENTITY_FILES) {
            break;
          }
          if (isFile(child)) {
            const childName = child.split(/[\\/]/).pop()!;
            if (
              childName.endsWith('.py') ||
              childName.endsWith('.ts') ||
              childName.endsWith('.js')
            ) {
              found.push(toPosix(relative(root, child)));
            }
          }
        }
      }
    }
  }
  return found.slice(0, MAX_ENTITY_FILES);
}

function* walk(root: string, maxDepth: number): IterableIterator<string> {
  const stack: Array<[string, number]> = [[root, 0]];
  while (stack.length > 0) {
    const [current, depth] = stack.pop()!;
    if (depth > maxDepth) {
      continue;
    }
    let children: string[];
    try {
      children = readdirSync(current).sort();
    } catch {
      continue;
    }
    for (const name of children) {
      if (IGNORED_DIRS.has(name)) {
        continue;
      }
      const child = join(current, name);
      if (isSymlink(child)) {
        continue;
      }
      yield child;
      if (isDirectory(child)) {
        stack.push([child, depth + 1]);
      }
    }
  }
}

function* walkAll(root: string): IterableIterator<string> {
  const stack: string[] = [root];
  while (stack.length > 0) {
    const current = stack.pop()!;
    let children: string[];
    try {
      children = readdirSync(current).sort();
    } catch {
      continue;
    }
    for (const name of children) {
      const child = join(current, name);
      yield child;
      if (isDirectory(child)) {
        stack.push(child);
      }
    }
  }
}

// posix path utility — unused export silences linters when the test
// suite doesn't import the helper directly.
void pathPosix;
